### PR TITLE
[SPARK-28141][SQL] Support special date values

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?

Supported special string values for `DATE` type. They are simply notational shorthands that will be converted to ordinary date values when read. The following string values are supported:
- `epoch [zoneId]` - `1970-01-01`
- `today [zoneId]` - the current date in the time zone specified by `spark.sql.session.timeZone`.
- `yesterday [zoneId]` - the current date -1
- `tomorrow [zoneId]` - the current date + 1
- `now` - the date of running the current query. It has the same notion as `today`.

For example:
```sql
spark-sql> SELECT date 'tomorrow' - date 'yesterday';
2
```

### Why are the changes needed?

To maintain feature parity with PostgreSQL, see [8.5.1.4. Special Values](https://www.postgresql.org/docs/12/datatype-datetime.html)

### Does this PR introduce any user-facing change?

Previously, the parser fails on the special values with the error:
```sql
spark-sql> select date 'today';
Error in query: 
Cannot parse the DATE value: today(line 1, pos 7)
```
After the changes, the special values are converted to appropriate dates:
```sql
spark-sql> select date 'today';
2019-09-06
```

### How was this patch tested?
- Added tests to `DateFormatterSuite` to check parsing special values from regular strings.
- Tests in `DateTimeUtilsSuite` check parsing those values from `UTF8String`
- Uncommented tests in `date.sql`

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala]: Here, there is the risk that `today` contains previous days if we were unlucky and the test was executed at the end of the day. Need to introduce some tolerance, I think.

**Comment by MaxGekk** [comment]
@dongjoon-hyun @maropu @cloud-fan Please, review the PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateFormatter.scala
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateFormatter.scala]: We always trim input inside of `stringToDate`. Maybe here we should trim the input in all cases? /cc @HyukjinKwon @srowen

**Comment by maropu** [comment]
The other DBMS-like systems support these syntaxes? If not, we should turn on/off these features by the compatibility mode flag https://issues.apache.org/jira/browse/SPARK-28934?

**Comment by maropu** [comment]
- How about the `timestamp` case? 
 - You dropped `inf/-inf`  in the supported?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateFormatter.scala
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateFormatter.scala]: Since casts from strings to date/tmestamp already does so, the trim looks reasonable to me.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: Just put the gen'd code for other reviewers;
```
/* 051 */         boolean project_isNull_0 = columnartorow_isNull_0;
/* 052 */         int project_value_0 = -1;
/* 053 */         if (!columnartorow_isNull_0) {
/* 054 */           scala.Option<Integer> project_intOpt_0 =
/* 055 */           org.apache.spark.sql.catalyst.util.DateTimeUtils.stringToDate(columnartorow_value_0, ((java.time.ZoneId) references[2] /* zoneId */));
/* 056 */           if (project_intOpt_0.isDefined()) {
/* 057 */             project_value_0 = ((Integer) project_intOpt_0.get()).intValue();
/* 058 */           } else {
/* 059 */             project_isNull_0 = true;
/* 060 */           }
/* 061 */         }
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: btw, (this is not related to this pr though), `scala.Option<Integer>` is ok for java compilers/jvm? @rednaxelafx I remember that jdk compilers cannot compile this statement: https://github.com/apache/spark/pull/21770#discussion_r202562989 because it seems the compilers erase the returned type from `scala.Option<Integer>` to `scala.Option<Object>`. I'm not sure why janino accepts this though....

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateFormatter.scala
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateFormatter.scala]: One thing I am a bit worried though, if users set the date `pattern` including the leading or trailing white spaces intentionally (because they know there's spaces), it will now returns invalid results.

```scala
DateFormatter(" yyyy MMM").parse("2018 Dec")
```
```
Text '2018 Dec' could not be parsed at index 0
java.time.format.DateTimeParseException: Text '2018 Dec' could not be parsed at index 0
	at java.base/java.time.format.DateTimeFormatter.parseResolved0(Unknown Source)
	at java.base/java.time.format.DateTimeFormatter.parse(Unknown Source)
	at java.base/java.time.LocalDate.parse(Unknown Source)
	at org.apache.spark.sql.catalyst.util.Iso8601DateFormatter.parse(DateFormatter.scala:36)
	at org.apache.spark.sql.util.DateFormatterSuite.$anonfun$new$15(DateFormatterSuite.scala:94)
	at scala.runtime.java8.JFunction0$mcI$sp.apply(JFunction0$mcI$sp.java:23)
```

While trimming makes sense in general, there's difference comparing to `stringToDate` since here we allow to set the `pattern` (and with this Java 8 APIs, now Spark can support exact and better match with the pattern).

Is this change necessary or does this change make it easier to fix the issue?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: super nit: add a blank line.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Better to avoid the `Try` use?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: This function always assumes the input has been trimmed?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: Could you add simple tests for this code path?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonGenerator.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonGenerator.scala]: ditto: add tests?

**Comment by MaxGekk** [comment]
@maropu Thank you for your review.
> You dropped inf/-inf in the supported?

I created the separate JIRA https://issues.apache.org/jira/browse/SPARK-29006 because the feature requires special date/timestamp values and handle them in all date/timestamp functions in the special way.

> How about the timestamp case?

Here is the JIRA https://issues.apache.org/jira/browse/SPARK-29012 and the PR: https://github.com/apache/spark/pull/25716 . I am still working on it because when I uncommented tests in `timestamp.sql`, it stated failing, see https://github.com/apache/spark/pull/25716#issuecomment-529013241

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Why?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Just for the codestyle? https://github.com/databricks/scala-style-guide#exception

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: I don't think the reasons from the style guide are applicable here. And code looks slightly ugly. Maybe you have better implementation than:
```scala
    def isValidZoneId(z: String): Boolean = {
      if (z == "") return true
      try {
        getZoneId(z)
        true
      } catch {
        case NonFatal(_) => false
      }
    }
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Added and renamed `specialDate` to `specialValue` since it is applicable to other types as well

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Changed the pattern to cut off blanks at the end. Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateFormatter.scala
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateFormatter.scala]: @HyukjinKwon I just asked because I am doing trimming here for checking special values. So, the overhead already exists. Maybe, it could be useful in another branch.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: I will revert it back because the input string is trimmed at both sides in any case.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: I added tests for the `from_csv` function to `CsvFunctionsSuite`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonGenerator.scala
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonGenerator.scala]: Added tests for `from_json`

**Comment by MaxGekk** [comment]
> The other DBMS-like systems support these syntaxes? 

@maropu I have not found other dbms that support such special values. 

> If not, we should turn on/off these features by the compatibility mode flag https://issues.apache.org/jira/browse/SPARK-28934?

The flag has not been merged yet: https://github.com/apache/spark/pull/25697

**Comment by maropu** [comment]
> If not, we should turn on/off these features by the compatibility mode flag https://issues.apache.org/jira/browse/SPARK-28934?
> The flag has not been merged yet: #25697

Yea, I meant we need to add the flag first? cc: @dongjoon-hyun @gengliangwang

**Comment by cloud-fan** [comment]
Yea I think we should have the flag first.

**Comment by maropu** [comment]
retest this please

**Comment by MaxGekk** [comment]
@maropu @dongjoon-hyun Here is the same question as in https://github.com/apache/spark/pull/25716#issuecomment-530697767

**Comment by MaxGekk** [comment]
I have rebased this on the master with merged https://github.com/apache/spark/pull/25716 . @HyukjinKwon Could you take a look at this, please.

**Comment by MaxGekk** [comment]
@dongjoon-hyun @maropu @HyukjinKwon Could take a look at this PR which is similar to already merged changes for the timestamp type.

**Review by maropu — APPROVED** [review]
Looks nice cc: @HyukjinKwon @dongjoon-hyun

**Comment by HyukjinKwon** [comment]
I am merging this just to unblock and easier to work further but please make sure to have a flag. Seems like that's pointed out by multiple committers such as @maropu and @cloud-fan.

**Comment by HyukjinKwon** [comment]
Otherwise, I will revert this and #25716

Merged to master.

**Comment by maropu** [comment]
We already have a jira ticket for linking this feature to the flag?

**Comment by MaxGekk** [comment]
This PR #25834 hides the feature under the SQL config spark.sql.dialect = "PostgreSQL"

**Comment by maropu** [comment]
Oh, I see. Thanks!

**Pull request closed without merging**
